### PR TITLE
[#51] [Frontend] As a user I can send a newsletter to a recipient

### PR DESCRIPTION
**Pull request by tednguyendev**
#51

## What happened 👀
- Add 2 new libs `classnames` and `react-multi-select-component`
- Add 1 new component [`MultiSelectWrapper`](https://github.com/nimblehq/js-framework-benchmark/pull/123/files#diff-242916bcb16bbd06bad4766b0f988a702d7fb74e140ff9338c2ccd8f66d63244R14) to use as a newsletter selector
- Add 1 new page [`SendNewsletter`](https://github.com/nimblehq/js-framework-benchmark/pull/123/files#diff-1410c2c61c8257ce65caa55dcf666a7b6315edc7ddd9c166d4b608a8fe552febR10) to contains the form to send newsletter to emails.
- Move logic to check protected routes from [`app/auth/sign-in/page.tsx`](https://github.com/nimblehq/js-framework-benchmark/pull/123/files#diff-e8b3129d5740bf43e4514d1c0d8e01bf9e8906c58721b29fb66dba4891d140d1L9) and [`app/page.tsx`](https://github.com/nimblehq/js-framework-benchmark/pull/123/files#diff-223e9b714cf2c9b5f163079f0bd63d3a6eac9c46915fc3afde20a2742897d706L62) to [`components/AuthLayout/index.tsx`](https://github.com/nimblehq/js-framework-benchmark/pull/123/files#diff-b894ae1556a3dc5ef73e274f940f27c8ec2194179fb627726ab078fab9f6ebe3R11) and [`components/AppLayout/index.tsx`](https://github.com/nimblehq/js-framework-benchmark/pull/123/files#diff-45d2814aa9e2f6cef0f484587e90fdb3161fcc6026a7816e1e4df7fdbd870ea9R15)
- Add tests and stylings

## Insight 📝
N/A

## Proof Of Work 📹

https://github.com/nimblehq/js-framework-benchmark/assets/76901736/4daca694-155d-4f39-9305-010de487dbbc

### CODE REVIEW

File: nextjs-13/src/components/AppLayout/index.tsx [hunk 1]
```diff
@@ -1,10 +1,48 @@
...
 import Header from '@components/Header';
 
 export default function Layout({ children }: { children: React.ReactNode }) {
+  const pathname = usePathname();
+  const isRoot = !pathname.includes('/send');
+  const { status } = useSession();
+
+  if (status === 'unauthenticated') {
+    redirect('/auth/sign-in');
+  }
```
Comment (olivierobert) [inline nextjs-13/src/components/AppLayout/index.tsx]: I noticed that this code was removed above https://github.com/nimblehq/js-framework-benchmark/pull/123/files#diff-e8b3129d5740bf43e4514d1c0d8e01bf9e8906c58721b29fb66dba4891d140d1. While this can be managed at the layout level, what do you think of using [the middleware](https://github.com/nimblehq/js-framework-benchmark/blob/develop/nextjs-13/src/middleware.ts) for that purpose? It would clean up component files.

File: nextjs-13/src/app/send/page.tsx [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    setSelected([]);
+    setEmail('');
+    getNewletters();
+  };
+
+  const handleSubmit = async (event) => {
+    event.preventDefault();
+
+    if (!selected.length) {
+      return makeToast('Please select atleast one newsletter', 'error');
```
Comment (olivierobert) [inline nextjs-13/src/app/send/page.tsx]: Minor typo:

```suggestion
      return makeToast('Please select at least one newsletter', 'error');
```

File: nextjs-13/src/app/send/page.tsx [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+  }, []);
+
+  const afterSubmit = () => {
+    setIsLoading(false);
+    setSelected([]);
+    setEmail('');
+    getNewletters();
+  };
+
+  const handleSubmit = async (event) => {
```
Comment (olivierobert) [inline nextjs-13/src/app/send/page.tsx]: What do you think of naming this:

```suggestion
  const resetState = async (event) => {
```

File: nextjs-13/src/components/AppLayout/index.test.tsx [hunk 4]
```diff
@@ -1,15 +1,67 @@
...
-  it('renders the component', () => {
-    render(
-      <AppLayout>
-        <></>
-      </AppLayout>
-    );
-
-    expect(screen.getByTestId('layout-default')).toBeVisible();
+  beforeEach(() => {
+    usePathname.mockReturnValue('localhost:3003/send');
```
Comment (olivierobert) [inline nextjs-13/src/components/AppLayout/index.test.tsx:14]: Why is this required?

File: nextjs-13/src/components/AppLayout/index.tsx [hunk 5]
```diff
@@ -1,10 +1,48 @@
+'use client';
+
+import classNames from 'classnames';
+import Link from 'next/link';
+import { redirect, usePathname } from 'next/navigation';
+import { useSession } from 'next-auth/react';
+
 import Header from '@components/Header';
 
 export default function Layout({ children }: { children: React.ReactNode }) {
```
Comment (olivierobert) [inline nextjs-13/src/components/AppLayout/index.tsx]: You might want to name it as per the file name to remove any confusion:

```suggestion
export default function AppLayout({ children }: { children: React.ReactNode }) {
```

Ditto below https://github.com/nimblehq/js-framework-benchmark/pull/123/files#diff-b894ae1556a3dc5ef73e274f940f27c8ec2194179fb627726ab078fab9f6ebe3L3

**Review by olivierobert — COMMENTED** [review]
Some minor and medium suggestions but overall it looks great 👍

File: nextjs-13/src/components/AppLayout/index.tsx (hunk 1 above)
Comment (tednguyendev) [reply nextjs-13/src/components/AppLayout/index.tsx]: At first, when I read the [NextAuth middleware page](https://next-auth.js.org/configuration/nextjs#middleware), I missed this [wrap middleware part](https://next-auth.js.org/configuration/nextjs#wrap-middleware), so I thought NextAuth only allow us to use middleware in a very [basic way](https://next-auth.js.org/configuration/nextjs#basic-usage) 👀 Let me try this wrap middleware way!

File: nextjs-13/src/app/send/page.tsx (hunk 3 above)
Comment (tednguyendev) [reply nextjs-13/src/app/send/page.tsx]: Yes I think that is a less vague name 😄

File: nextjs-13/src/components/AppLayout/index.test.tsx (hunk 4 above)
Comment (tednguyendev) [reply nextjs-13/src/components/AppLayout/index.test.tsx:14]: Because I'm using `usePathname ` [here](https://github.com/nimblehq/js-framework-benchmark/pull/123/files#diff-45d2814aa9e2f6cef0f484587e90fdb3161fcc6026a7816e1e4df7fdbd870ea9R11-R12) to get the path name to [differentiate the background-color of the selected tabs and non-selected tabs](https://github.com/nimblehq/js-framework-benchmark/pull/123/files#diff-45d2814aa9e2f6cef0f484587e90fdb3161fcc6026a7816e1e4df7fdbd870ea9R27) 👀 

CleanShot 2023-06-21 at 13 14 45@2x

CleanShot 2023-06-21 at 13 15 21@2x

File: nextjs-13/src/components/AppLayout/index.tsx (hunk 5 above)
Comment (tednguyendev) [reply nextjs-13/src/components/AppLayout/index.tsx]: Resolved [here](https://github.com/nimblehq/js-framework-benchmark/pull/123/commits/3b1a083a4265078a2f9995befb15bf3176a4d851)

File: nextjs-13/src/app/send/page.tsx (hunk 3 above)
Comment (tednguyendev) [reply nextjs-13/src/app/send/page.tsx]: Resolved [here](https://github.com/nimblehq/js-framework-benchmark/pull/123/commits/126912f87b99d22d6544794aba9649f540bae17a)

File: nextjs-13/src/app/send/page.tsx (hunk 2 above)
Comment (tednguyendev) [reply nextjs-13/src/app/send/page.tsx]: Resolved [here](https://github.com/nimblehq/js-framework-benchmark/pull/123/files#diff-1410c2c61c8257ce65caa55dcf666a7b6315edc7ddd9c166d4b608a8fe552febR45) and [here](https://github.com/nimblehq/js-framework-benchmark/pull/123/files#diff-79f2f429dd86ce329745b8c7cd2baf8ffbaa9bc8541c610395d99a79eb0dfea5R109)

File: nextjs-13/src/components/AppLayout/index.tsx (hunk 1 above)
Comment (tednguyendev) [reply nextjs-13/src/components/AppLayout/index.tsx]: Turned out that that wrap middleware is not ideal also 🤣, so I decided to just use the normal next.js middleware way. Resolved [here](https://github.com/nimblehq/js-framework-benchmark/pull/123/commits/0e71f45a42ac09949e190ab90a347df4d9994e2f#diff-8dc469b9cf60785ce08e724cab36052f5f6278de35da8aea179287ddbe5c6ac3R4-R19)

Comment (olivierobert) [reply nextjs-13/src/components/AppLayout/index.tsx]: Next.js middleware's capabilities are very limited indeed (there can be only one 🤯 ), but it usually suffices for simple applications. The improvement looks good.

File: nextjs-13/src/components/AppLayout/index.test.tsx (hunk 4 above)
Comment (olivierobert) [reply nextjs-13/src/components/AppLayout/index.test.tsx:14]: So you should add a comment above the code as it is quite unexpected

File: nextjs-13/src/app/auth/sign-in/page.test.tsx [hunk 6]
```diff
@@ -1,40 +1,30 @@
...
 jest.mock('next/navigation');
 
 describe('SignInPage', () => {
-  describe('Session status is "authenticated', () => {
-    it('redirects to home page', () => {
-      useSession.mockReturnValue({ status: 'authenticated' });
-      render(<SignInPage />);
-      expect(redirect).toHaveBeenCalledWith('/');
-    });
+  it('renders h4', () => {
```
Comment (olivierobert) [inline nextjs-13/src/app/auth/sign-in/page.test.tsx]: It should be the following to be consistent with the other tests:

```suggestion
  it('renders the title', () => {
```

File: nextjs-13/src/app/page.test.tsx [hunk 7]
```diff
@@ -22,92 +20,66 @@ describe('Home', () => {
...
-      await waitFor(() => {
-        expect(redirect).toHaveBeenCalledWith('/auth/sign-in');
-      });
+    await waitFor(() => {
+      expect(screen.getByText('Your Newsletters')).toBeInTheDocument();
     });
   });
 
-  describe('Session status is "authenticated', () => {
+  describe('giving fetching data', () => {
```
Comment (olivierobert) [inline nextjs-13/src/app/page.test.tsx]: I missed this incorrect phrasing:

```suggestion
  describe('given fetching data', () => {
```

**Review by olivierobert — CHANGES_REQUESTED** [review]
Just very tiny issues to fix then this can be merged

File: nextjs-13/src/components/AppLayout/index.test.tsx (hunk 4 above)
Comment (tednguyendev) [reply nextjs-13/src/components/AppLayout/index.test.tsx:14]: Resolved [here](https://github.com/nimblehq/js-framework-benchmark/pull/123/commits/6a509538995f3fa7b0bcd282e2f6ebcbb6e6466e) 👀

File: nextjs-13/src/app/auth/sign-in/page.test.tsx (hunk 6 above)
Comment (tednguyendev) [reply nextjs-13/src/app/auth/sign-in/page.test.tsx]: Fixed [here](https://github.com/nimblehq/js-framework-benchmark/pull/123/commits/108cc7cf1e36665071c8d25ff89f06df610ed39c#diff-c54ff5d2e911b6e1771914307c60d6c0ca79b7f9d835e08eac133fdf4c7e98cbR11)

File: nextjs-13/src/app/page.test.tsx (hunk 7 above)
Comment (tednguyendev) [reply nextjs-13/src/app/page.test.tsx]: Fixed [here](https://github.com/nimblehq/js-framework-benchmark/pull/123/commits/108cc7cf1e36665071c8d25ff89f06df610ed39c#diff-4b9ae30a4e34a98e0e0d12af3e4872e7ac9865c68f016c1e6667d4336ce08ce1R31-R51)

**Review by olivierobert — APPROVED** [review]
